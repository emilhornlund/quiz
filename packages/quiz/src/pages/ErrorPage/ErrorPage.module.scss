@use '../../styles/colors';
@use '../../styles/helpers';

.errorIcon {
  @include helpers.fontSize(148px);

  color: colors.$red-1;
}