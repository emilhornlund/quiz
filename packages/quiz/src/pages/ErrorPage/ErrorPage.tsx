import { faHeartCrack } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import React, { FC } from 'react'
import { useNavigate, useRouteError } from 'react-router-dom'

import { IconButtonArrowLeft, Page, Typography } from '../../components'

import styles from './ErrorPage.module.scss'

const ErrorPage: FC = () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const error: any = useRouteError()
  const navigate = useNavigate()

  return (
    <Page
      header={
        <IconButtonArrowLeft
          id="go-back"
          type="button"
          size="small"
          value="Go Back"
          onClick={() => {
            navigate(-1)
          }}
        />
      }>
      <div className={styles.errorIcon}>
        <FontAwesomeIcon icon={faHeartCrack} />
      </div>
      <Typography variant="title" size="medium">
        Oops! Something went wrong.
      </Typography>
      <Typography variant="subtitle" size="medium">
        {error?.statusText || 'An unexpected error occurred.'}
      </Typography>
      <Typography variant="text" size="medium">
        {error?.data ||
          'Please try refreshing the page or go back to the homepage.'}
      </Typography>
    </Page>
  )
}

export default ErrorPage
