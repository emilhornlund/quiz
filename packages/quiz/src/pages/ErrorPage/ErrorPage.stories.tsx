import type { Meta, StoryObj } from '@storybook/react'
import { withRouter } from 'storybook-addon-remix-react-router'

import ErrorPage from './ErrorPage'

const meta = {
  component: ErrorPage,
  decorators: [withRouter],
  parameters: {
    layout: 'fullscreen',
  },
} satisfies Meta<typeof ErrorPage>

export default meta
type Story = StoryObj<typeof meta>

export const Default = {
  args: {},
} satisfies Story
