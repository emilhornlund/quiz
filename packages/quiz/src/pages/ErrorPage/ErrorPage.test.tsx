import { render, screen } from '@testing-library/react'
import React from 'react'
import { MemoryRouter, useNavigate, useRouteError } from 'react-router-dom'
import { describe, expect, it, vi } from 'vitest'

import ErrorPage from './ErrorPage'

// Mock `useRouteError` and `useNavigate` to simulate error responses and navigation
vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom')
  return {
    ...actual,
    useRouteError: vi.fn(),
    useNavigate: vi.fn(), // Define useNavigate as a mock function
  }
})

describe('ErrorPage', () => {
  it('should render the default error message', () => {
    ;(useRouteError as jest.Mock).mockReturnValueOnce(null)

    render(
      <MemoryRouter>
        <ErrorPage />
      </MemoryRouter>,
    )

    expect(screen.getByText('Oops! Something went wrong.')).toBeInTheDocument()
    expect(
      screen.getByText('An unexpected error occurred.'),
    ).toBeInTheDocument()
  })

  it('should display custom error message based on route error', () => {
    ;(useRouteError as jest.Mock).mockReturnValueOnce({
      statusText: 'Not Found',
      data: 'The requested page could not be found.',
    })

    render(
      <MemoryRouter>
        <ErrorPage />
      </MemoryRouter>,
    )

    expect(screen.getByText('Not Found')).toBeInTheDocument()
    expect(
      screen.getByText('The requested page could not be found.'),
    ).toBeInTheDocument()
  })

  it('should navigate back when the back button is clicked', () => {
    const navigateMock = vi.fn()
    ;(useNavigate as jest.Mock).mockReturnValue(navigateMock)

    render(
      <MemoryRouter>
        <ErrorPage />
      </MemoryRouter>,
    )

    const backButton = screen.getByRole('button', { name: /go back/i })
    backButton.click()

    expect(navigateMock).toHaveBeenCalledWith(-1)
  })

  it('should match snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <ErrorPage />
      </MemoryRouter>,
    )

    expect(container).toMatchSnapshot()
  })
})
