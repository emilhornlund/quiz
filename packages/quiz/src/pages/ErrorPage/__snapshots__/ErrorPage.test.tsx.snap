// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorPage > should match snapshot 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <div
        class="logo"
      >
        <span
          class="icon"
        />
        <span
          class="text"
        >
          Quiz
        </span>
      </div>
      <div
        class="side"
      >
        <div
          class="main small"
        >
          <button
            class="primary"
            data-testid="test-go-back-button"
            id="go-back"
            name="go-back"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-left "
              data-icon="arrow-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.2 288 416 288c17.7 0 32-14.3 32-32s-14.3-32-32-32l-306.7 0L214.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
                fill="currentColor"
              />
            </svg>
            <span>
              Go Back
            </span>
          </button>
        </div>
        <div
          class="avatar"
        >
          <button
            type="button"
          >
            <img
              alt="Profile"
              src="/src/assets/images/avatar.svg"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="content fullWidth centerAlign"
    >
      <div
        class="errorIcon"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-heart-crack "
          data-icon="heart-crack"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M119.4 44.1c23.3-3.9 46.8-1.9 68.6 5.3l49.8 77.5-75.4 75.4c-1.5 1.5-2.4 3.6-2.3 5.8s1 4.2 2.6 5.7l112 104c2.9 2.7 7.4 2.9 10.5 .3s3.8-7 1.7-10.4l-60.4-98.1 90.7-75.6c2.6-2.1 3.5-5.7 2.4-8.8L296.8 61.8c28.5-16.7 62.4-23.2 95.7-17.6C461.5 55.6 512 115.2 512 185.1l0 5.8c0 41.5-17.2 81.2-47.6 109.5L283.7 469.1c-7.5 7-17.4 10.9-27.7 10.9s-20.2-3.9-27.7-10.9L47.6 300.4C17.2 272.1 0 232.4 0 190.9l0-5.8c0-69.9 50.5-129.5 119.4-141z"
            fill="currentColor"
          />
        </svg>
      </div>
      <div
        class="typography title medium"
      >
        Oops! Something went wrong.
      </div>
      <div
        class="typography subtitle medium"
      >
        An unexpected error occurred.
      </div>
      <div
        class="typography text medium"
      >
        Please try refreshing the page or go back to the homepage.
      </div>
    </div>
  </div>
</div>
`;
